include/dt-bindings/gpio/gpio.h

```diff
@@ -6,107 +6,157 @@
 #ifndef ZEPHYR_INCLUDE_DT_BINDINGS_GPIO_GPIO_H_
 #define ZEPHYR_INCLUDE_DT_BINDINGS_GPIO_GPIO_H_
 
-
 /**
  * @name GPIO direction flags
- * The `GPIO_DIR_*` flags are used with `gpio_pin_configure` or `gpio_port_configure`,
- * to specify whether a GPIO pin will be used for input or output.
+ * The `GPIO_DIR_*` flags are used to specify whether a GPIO pin will be used
+ * for input and/or output and the initial state for output.
  * @{
  */
-/** GPIO pin to be input. */
-#define GPIO_DIR_IN		(0 << 0)
-
-/** GPIO pin to be output. */
-#define GPIO_DIR_OUT		(1 << 0)
-
 /** @cond INTERNAL_HIDDEN */
-#define GPIO_DIR_MASK		0x1
+#define GPIO_DIR_SHIFT          0
+#define GPIO_DIR_MASK		(0x7 << GPIO_DIR_SHIFT)
 /** @endcond */
-/** @} */
-
 
-/**
- * @name GPIO interrupt flags
- * The `GPIO_INT_*` flags are used with `gpio_pin_configure` or `gpio_port_configure`,
- * to specify how input GPIO pins will trigger interrupts.
- * @{
- */
-/** GPIO pin to trigger interrupt. */
-#define GPIO_INT		(1 << 1)
-
-/** GPIO pin trigger on level low or falling edge. */
-#define GPIO_INT_ACTIVE_LOW	(0 << 2)
-
-/** GPIO pin trigger on level high or rising edge. */
-#define GPIO_INT_ACTIVE_HIGH	(1 << 2)
+/** Configures GPIO pin as input only. */
+#define GPIO_DIR_IN             (0 << GPIO_DIR_SHIFT)
+/** Configures GPIO pin as output with no change to output state. */
+#define GPIO_DIR_OUT            (1 << GPIO_DIR_SHIFT)
+/** Configures GPIO pin as output low. */
+#define GPIO_DIR_OUT_LOW        (2 << GPIO_DIR_SHIFT)
+/** Configures GPIO pin as output high. */
+#define GPIO_DIR_OUT_HIGH       (3 << GPIO_DIR_SHIFT)
+/** Configures GPIO pin as disabled / disconnected. */
+#define GPIO_DIR_DISABLED       (4 << GPIO_DIR_SHIFT)
```

Comment:
This doesn't seem right.  In #11880 the base conditions are:
* `GPIO_IO_IN_ENABLED` enables or disables input from the GPIO.  On some devices this corresponds to connecting or disconnecting the input buffer.
* `GPIO_IO_OUT_ENABLED` enables or disables output from the GPIO.

I think the equivalent here would be:
* neither input nor output: `GPIO_DIR_DISABLED`
* input-only: `GPIO_DIR_IN` (== 0)
* output-only: not expressible?
* input and output: `GPIO_DIR_OUT | GPIO_DIR_IN` == `GPIO_DIR_OUT`.

So this fails to provide desirable functionality.

Reply:
>     * output-only: not expressible?

This is indeed not supported. There is only so much one can do by extending GPIO_DIR_ flags. To have it added we would need to deprecate GPIO_DIR_ flags and provide new configuration based on pinctrl interface which does it in a clean and consistent way.

Reply:
I thought I managed to do it.  Though that was only because existing code uses:
```
               if ((flags & GPIO_DIR_MASK) == GPIO_DIR_IN) {
```
taking advantage of the power of masks for single-bit fields....

Reply:
I've reworked the flags to allow independently configure input/output mode. This is consistent with the approach used by Linux pinctr/gpio drivers. The output only mode is now possible.